### PR TITLE
Resurrect electron shell

### PR DESCRIPTION
**Pull request by frantic**
This reverts commit 3771a1e54aed9121cb0e0aeea130fd01850682ee and brings it up to date with recent devtools changes.

Test Plan:
```
npm run lint
flow
webpack
electron .
```
Opened RN app, made sure it connects. Seems like there are still a couple of edge case bugs and sometimes I need to reload either devtools or react app. Will work on tests for those later.

@jaredly @spicyj

### CODE REVIEW

**Comment by spicyj** [comment]
What did you change? This would be easier to review as two commits (one straight revert, one with fixes).

**Comment by jaredly** [comment]
I've done some more work on the electron shell in [this branch here](https://github.com/facebook/react-devtools/compare/master...jaredly:inspection-server?expand=1). It also allows you to inspect chrome apps and the like (re #91)

It's mostly the same as this PR, but with some more organization & misc fixes, as well as the "inspect chrome apps / webviews" stuff.

**Comment by frantic** [comment]
@spicyj @jaredly what do you guys think about the approach in last 4 commits? I'm trying to isolate the electron UI into self-contained package that can be re-used in Atom/Nuclide and other Electron-based apps.

`shells/electron/node/index.js` still needs some cleanup and customization options (like port to listen, etc.)

@jaredly I like the approach you take with serving `embed.js` file, I'm happy to integrate it here too. In RN we could easily change `setupDevtools` to `eval(fetch('/embed'))`.

**Comment by jaredly** [comment]
I like it

**Comment by darrrk** [comment]
Any moves here?

**Comment by matthewwithanm** [comment]
I think `run.js` needs to be updated to setup the Relay backend too. Otherwise, the `relay:check` call triggers a yellow box ("unknown call").

File: shells/electron/src/backend.js [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree. An additional grant
+ * of patent rights can be found in the PATENTS file in the same directory.
+ *
+ * @flow
+ */
+
+'use strict';
+
+declare var FOR_BACKEND: {
```
Comment (spicyj) [inline shells/electron/src/backend.js]: Where does this come from? I don't see this being set anywhere.

File: shells/electron/src/backend.js [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+  wall: Object;
+  resolveRNStyle: () => void;
+};
+
+if (!window.performance) {
```
Comment (spicyj) [inline shells/electron/src/backend.js]: Doesn't electron provide this?

**Comment by spicyj** [comment]
I don't have all the pieces in my head to review this properly, but let's just get this in.

File: shells/electron/package.json [hunk 3]
```diff
@@ -0,0 +1,9 @@
+{
+  "main": "./index.js",
+  "scripts": {
+    "build": "webpack --config webpack.backend.js && webpack --config webpack.config.js"
```
Comment (iamdustan) [inline shells/electron/package.json]: what does this expect your path to be? Or is webpack expected to be installed globally?

File: shells/electron/package.json [hunk 4]
```diff
@@ -0,0 +1,9 @@
...
+  },
+  "dependencies": {
+    "ws": "^0.7.2"
```
Comment (iamdustan) [inline shells/electron/package.json]: why 0.7.2? 1.1.0 is latest.

0.7.2 seems like this requires an older version of node. Should that be added to the engines field so consumers know of that requirement?

Comment (frantic) [inline shells/electron/package.json]: `1.1.0` is great (and doesn't require compiling C++ modules)! This PR is rather old, that's why it had `0.7.2`

File: shells/electron/src/backend.js (hunk 2 above)
Comment (iamdustan) [inline shells/electron/src/backend.js]: I removed it locally and everything was still okay. Running latest electron-prebuilt.

File: shells/electron/src/ui.js [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+ *
+ */
+'use strict';
+
+var ws = require('ws');
+
+var installGlobalHook = require('../../../backend/installGlobalHook');
+installGlobalHook(window);
+var Panel = require('../../../frontend/Panel');
+var React = require('react');
```
Comment (iamdustan) [inline shells/electron/src/ui.js]: s/react/react-dom ?

File: shells/electron/src/backend.js (hunk 2 above)
Comment (frantic) [inline shells/electron/src/backend.js]: This code is injected and `eval`ed inside another context that might not have `windows.performance`

File: shells/electron/src/backend.js (hunk 1 above)
Comment (frantic) [inline shells/electron/src/backend.js]: Will add link to https://github.com/facebook/react-native/blob/b00c77af8066cf74f05ccaca2b08c8952e1ae8a6/Libraries/Devtools/setupDevtools.js#L19

File: shells/electron/package.json (hunk 3 above)
Comment (frantic) [inline shells/electron/package.json]: I think we can re-use `webpack` that was installed via root `react-devtools/package.json`

File: shells/electron/src/backend.js (hunk 2 above)
Comment (iamdustan) [inline shells/electron/src/backend.js]: ah yes. I copied that for here: https://github.com/iamdustan/react-hardware/blob/516d2a8854028577a7d7928729042c8738181bb0/src/devtools/setupDevtools.js

File: shells/electron/package.json (hunk 3 above)
Comment (iamdustan) [inline shells/electron/package.json]: yeah. What you changed it to was what I had to run locally to reuse that webpack. :thumbsup:

**Comment by iamdustan** [comment]
🎉
